[Feature Request] Add detailed git stats for a specific branch

joshuadeguzman opened the pull request:
### Summary
This feature request is very straightforward, it's for us to track updates from the other branch (especially on the latest `master` or repos with `other-major-feature` branches) from time to time without switching branches so it'll save up time not re-running file indexing in IDEs, stop the current feature we are working on, etc.

### Tasks
- [X] Add non-interactive detailed git stats for a specific branch
- [X] Update `OPTIONS` list
- [X] Update and pass unit test

### Test Plan
**Access via interactive menu**
<img width="526" alt="Screen Shot 2019-05-10 at 12 26 06 AM" src="https://user-images.githubusercontent.com/20706361/57469884-3d556180-72ba-11e9-8a42-83161fcacf0c.png">
> **Expected Results**: Detailed git stats should still work fine as it was before

**Access via non-interactive menu happy path**
<img width="638" alt="Screen Shot 2019-05-10 at 12 25 39 AM" src="https://user-images.githubusercontent.com/20706361/57469890-3f1f2500-72ba-11e9-8dba-4603bfd41566.png">
> **Expected Results**: Data should display the stats within the specified branch

**Access via non-interactive menu edge case**
<img width="637" alt="Screen Shot 2019-05-10 at 12 25 52 AM" src="https://user-images.githubusercontent.com/20706361/57469893-3f1f2500-72ba-11e9-9ec0-4519172408ef.png">
> **Expected Results**: If specified branch does not exist, data should display the stats within the current branch

joshuadeguzman commented:
This also, failed build was caused by the travis configuration. Let me know your thoughts.

tomice commented:
On lines 175 and 179: Prefer bash's built-in [[ to the traditional [ test. It allows for string matching and regular expressions should we need them in the future.

On line 178 in git-quick-stats: Please prefer $(..) to `..` as `..` has undefined behaviors in relation to quoting in POSIX, is harder to nest than $(..), and may bring about some strange results when you attempt to escape it.

On line 178 in git-quick-stats: Make sure ${branch} is quoted to prevent globbing and word splitting issues.

On line 178: does_exist is a global variable in this instance. However, it is never used outside of the scope of detailedGitStats from what I can see. Prefer declaring this as a local var.

On line 181: Prefer double quoting ${branch} to help prevent possible globbing issues.

For the entire logic block of "Check if requesting for a specific branch" section, prefer calling the command directly instead of storing the command's result as a non-empty var. For example, I believe this code will do what you're attempting to do (please double-check it, though):

```
if [[ -n "${branch}" ]]; then
    if git show-ref refs/heads/"${branch}"; then
        is_branch_existing=true
        _branch="${branch}"
    else
        is_branch_existing=false
        _branch=""
    fi
fi
```

joshuadeguzman commented:
Will check this out, learned a lot from this PR! Thank you! 🎉

joshuadeguzman commented:
Hi @tomice, I have updated my diff. Thank you for helpful insights!

tomice commented:
Very nice :) LGTM. I think this would be a useful feature to add. Two more questions, if I may...

1) During the edge case of no branch existing when queried for a branch, the bold red text output says "Contribution stats (by author)" and proceeds to list git information. This could possibly be confusing to users. Do you think it would make sense to, instead, always list the branch so that the user knows exactly where this info is coming from?

2) Currently, this feature works via non-interactive mode. Thoughts about adding this as an option to the interactive menu, as well?

joshuadeguzman commented:
You're welcome!

1. I am thinking of updating it to "Contribution stats...on the current branch:" or maybe be display the name of the current branch.
2. I don't mind creating one for the interactive menu. Should I just include that in this PR and retitle this?

Let me know what works for you guys :)

tomice commented:
Either the "on current branch" suggestion or displaying the branch's name would work great. I'll let you choose which you think is best. :) And I think changing the title of this PR and including the interactive option in here would be the easiest. This way, we'd get both the interactive and non-interactive options in one PR instead of having them split between two.

Thanks again for a very well-written PR and your overall contribution to this project. It is greatly appreciated :)

joshuadeguzman commented:
### Summary 
- [X] Added interactive menu
- [X] Optimize checking of branch existence
- [X] Add "on the current branch"

> I chose "on the current branch" rather than retrieving the current branch name, because it will just make the runtime and logic a little bit complex to execute.

### Updated Test Plan
This includes the inclusion of the feature to the interactive menu

**Interactive menu (branch not existing)**
![Screen Shot 2019-05-10 at 6 55 34 PM](https://user-images.githubusercontent.com/20706361/57522681-768edf80-7355-11e9-81a5-96bce0fcb46d.png)

**Interactive menu (branch existing)**
![Screen Shot 2019-05-10 at 6 56 04 PM](https://user-images.githubusercontent.com/20706361/57522685-7858a300-7355-11e9-8f28-c9853b9e2277.png)

**Non-interactive menu (branch not existing)**
![Screen Shot 2019-05-10 at 6 56 36 PM](https://user-images.githubusercontent.com/20706361/57522737-8f979080-7355-11e9-8fdd-e68b4d52d064.png)

**Non-interactive menu (branch existing)**
![Screen Shot 2019-05-10 at 6 56 24 PM](https://user-images.githubusercontent.com/20706361/57522742-91615400-7355-11e9-97bf-e0cd3cf52fba.png)

tomice commented:
Looks great! Don't forget that we will want to update the manpage (git-quick-stats.1) and documentation (README.md) to reflect these new changes. Otherwise, we will be adding undocumented functionality.

joshuadeguzman commented:
### Updates
- [X] Updated README.md
- [X] Updated manpage

### Attachments

**Addition of `--git-stats-by-branch` in the manpage**
<img width="456" alt="Screen Shot 2019-05-10 at 11 58 40 PM" src="https://user-images.githubusercontent.com/20706361/57540622-afdc4500-737f-11e9-984a-35e777a26ee3.png">

joshuadeguzman commented on git-quick-stats.1:
```diff
@@ -24,12 +24,16 @@ This program allows you to see detailed information about a git repository.
 show the best people to contact to review code
 .HP
 .PP
-.PP
```
Had to remove this for a much better consistency.
